File: src/services/state.js
import { promises as fs } from 'fs';
import path from 'path';
import { fileURLToPath } from 'url';
import logger from '../utils/logger.js';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const STATE_FILE = path.resolve(__dirname, '../../.bot_state.json');

export async function loadState() {
  try {
    const raw = await fs.readFile(STATE_FILE, 'utf8');
    return JSON.parse(raw);
  } catch (err) {
    if (err.code === 'ENOENT') {
      logger.debug('State file does not exist, returning empty state');
      return {};
    }
    logger.error('Failed to load state file', { error: err.message });
    return {};
  }
}

export async function saveState(state) {
  try {
    await fs.writeFile(STATE_FILE, JSON.stringify(state, null, 2), 'utf8');
    logger.debug('State file saved');
  } catch (err) {
    logger.error('Failed to save state file', { error: err.message });
    throw err;
  }
}

export async function getStateValue(key) {
  const state = await loadState();
  return state[key];
}

export async function setStateValue(key, value) {
  const state = await loadState();
  state[key] = value;
  await saveState(state);
}

export async function deleteStateValue(key) {
  const state = await loadState();
  delete state[key];
  await saveState(state);
}

export default {
  loadState,
  saveState,
  getStateValue,
  setStateValue,
  deleteStateValue,
};

File: src/services/state.py
"""Bot state persistence module."""
import json
from pathlib import Path
from src.utils.logger import logger


STATE_FILE = Path(__file__).parent.parent.parent / '.bot_state.json'


async def load_state() -> dict:
    """Load state from file."""
    try:
        if STATE_FILE.exists():
            with open(STATE_FILE, 'r') as f:
                return json.load(f)
        logger.debug('State file does not exist, returning empty state')
        return {}
    except Exception as err:
        logger.error('Failed to load state file', {'error': str(err)})
        return {}


async def save_state(state: dict):
    """Save state to file."""
    try:
        with open(STATE_FILE, 'w') as f:
            json.dump(state, f, indent=2)
        logger.debug('State file saved')
    except Exception as err:
        logger.error('Failed to save state file', {'error': str(err)})
        raise


async def get_state_value(key: str):
    """Get a single state value."""
    state = await load_state()
    return state.get(key)


async def set_state_value(key: str, value):
    """Set a single state value."""
    state = await load_state()
    state[key] = value
    await save_state(state)


async def delete_state_value(key: str):
    """Delete a single state value."""
    state = await load_state()
    if key in state:
        del state[key]
        await save_state(state)
